daemons/controld/controld_join_dc.c

```diff
@@ -29,6 +41,72 @@ gboolean check_join_state(enum crmd_fsa_state cur_state, const char *source);
  */
 static int current_join_id = 0;
 
+/*!
+ * \internal
+ * \brief Destroy the hash table containing failed sync nodes
+ */
+void
+controld_destroy_failed_sync_table(void)
+{
+    if (failed_sync_nodes != NULL) {
+        g_hash_table_destroy(failed_sync_nodes);
+        failed_sync_nodes = NULL;
+    }
+}
+
+/*!
+ * \internal
+ * \brief Remove a node from the failed sync nodes table if present
+ *
+ * \param[in] node_name  Node name to remove
+ */
+void
+controld_remove_failed_sync_node(const char *node_name)
+{
+    if (failed_sync_nodes != NULL) {
+        g_hash_table_remove(failed_sync_nodes, (gchar *) node_name);
+    }
+}
+
+/*!
+ * \internal
+ * \brief Add to a hash table a node whose CIB failed to sync
+ *
+ * \param[in] node_name  Name of node whose CIB failed to sync
+ * \param[in] join_id    Join round when the failure occurred
+ */
+static void
+record_failed_sync_node(const char *node_name, int join_id)
+{
+    if (failed_sync_nodes == NULL) {
+        failed_sync_nodes = pcmk__strikey_table(g_free, NULL);
+    }
+
+    /* Assert if the node is already in the table. That would indicate that we
+     * failed to nack it during the filter offer step.
+     */
+    assert(g_hash_table_insert(failed_sync_nodes, g_strdup(node_name),
+                               (gpointer *) &join_id));
```

Comment:
I'm realizing this is probably unsafe because the memory will be freed for `join_id`, and I'm wondering if I have a similar mistake in the "don't duplicate failcounts" PR even though test runs succeeded there. Probably need to allocate a new int and give ownership to the hash table.

Reply:
As mentioned elsewhere, you can use the GPOINTER_TO_INT() etc. macros to avoid memory allocation